### PR TITLE
CSS line numbers for logs

### PR DESCRIPTION
**Pull request by benmandrew**
Fix for #654 using @novemberkilo's CSS suggestion.

Copying logs no longer pulls in line numbers (tested in Safari 15.2).

### CODE REVIEW

**Comment by benmandrew** [comment]
Not yet ready for merging: multi-line log lines do not have line numbers or associated formatting beyond the first line (see image), and causes the logs background issue in #716 to return.

Screenshot 2023-01-10 at 13 21 29

**Comment by maiste** [comment]
Thanks for the PR :+1: 

I've tested _Copy/Paste_ for `Step to reproduce` and `Logs` on:
- Mozilla Firefox 108.0.1
- Chromium 108.0.5359.124

It seems to work fine (appart for the issue you where suggesting). The background problem can be solved in this configuration by adding `bg-default` to this [line](https://github.com/ocurrent/ocaml-ci/blob/108990a014bf7452c0cfbbc4a13becf5cdff3b92/web-ui/view/step.ml#L423).

File: web-ui/static/css/style.css [hunk 1]
```diff
@@ -206,34 +206,69 @@ select {
...
+  user-select: none;
+  -moz-user-select: none;
+  -webkit-user-select: none;
+}
+
+pre.code span.th::before {
+  content: counter(linenum);
+  text-align: right;
+  display: block;
   @apply text-gray-900 dark:text-gray-100 bg-gray-200 dark:bg-gray-850 text-center px-1 md:px-6 w-14 text-gray-700 text-sm cursor-pointer select-none border-none;
```
Comment (maiste) [inline web-ui/static/css/style.css:239]:
```suggestion
  content: counter(linenum);
  @apply block text-gray-900 dark:text-gray-100 bg-gray-200 dark:bg-gray-850 text-center px-1 md:px-6 w-14 text-gray-700 text-sm text-right cursor-pointer select-none border-none;
```

File: web-ui/static/css/style.css [hunk 2]
```diff
@@ -206,34 +206,69 @@ select {
...
+pre.code::before {
+  counter-reset: linenum;
 }
 
-.code-line:last-child .code-line__number,
-.code-line:last-child .code-line__code {
-  @apply pb-3;
+pre.code span.tr {
+  display: table-row;
+  counter-increment: linenum;
```
Comment (maiste) [inline web-ui/static/css/style.css:225]:
```suggestion
  counter-increment: linenum;
    @apply table-row;
```

File: web-ui/static/css/style.css [hunk 3]
```diff
@@ -206,34 +206,69 @@ select {
...
 }
 
+pre.code {
+  display: table;
+  table-layout: fixed;
+  width: 100%;
+  white-space: pre-wrap;
+  @apply whitespace-pre border-none text-sm;
+  font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
+  border-left: 1px solid transparent;
```
Comment (maiste) [inline web-ui/static/css/style.css:216]:
```suggestion
  @apply table table-fixed w-full whitespace-pre-wrap border-l  border-solid border-transparent text-sm;
  font-family: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace;
```

**Review by maiste — COMMENTED** [review]
Now that we have the tailwind machinery setup in place, it's nice to use it to avoid having conflict between custom CSS and tailwind generated one (e.g. `whitespace` as an example). It's also capable of optimising some parts of the generated code, so we should use it as much as possible :)

File: web-ui/view/step.ml [hunk 4]
```diff
@@ -620,14 +620,32 @@ module Make (M : Git_forge_intf.Forge) = struct
       in
       List.filter_map aux data |> String.concat "\n"
     in
+    let collapse_carriage_returns log_line =
+      let rec last = function
+        | [] -> raise (Failure "Trying to take log_line from empty list (BUG)")
+        | s :: [] -> s
+        | _ :: l -> last l
```
Comment (maiste) [inline web-ui/view/step.ml:627]:
```suggestion
        | [ s ] -> s
        | _ :: l -> last l
```

File: web-ui/view/step.ml [hunk 5]
```diff
@@ -620,14 +620,32 @@ module Make (M : Git_forge_intf.Forge) = struct
...
+      | log_line ->
+          Dream.log "%S" log_line;
+          let split = Astring.String.cuts ~sep:"\r" log_line in
+          if List.length split > 1 then List.iter (Dream.log "%S") split;
+          last split
+    in
+    let process_logs data =
+      data
+      |> Astring.String.cuts ~sep:"\n"
+      |> List.map (fun l -> l |> collapse_carriage_returns |> Ansi.process ansi)
```
Comment (maiste) [inline web-ui/view/step.ml:640]:
```suggestion
      |> List.map (fun l -> collapse_carriage_returns l |> Ansi.process ansi)
```

File: web-ui/view/step.ml [hunk 6]
```diff
@@ -551,10 +566,11 @@ module Make (M : Git_forge_intf.Forge) = struct
     let line_number = ref 0 in
     let last_line_blank = ref false in
     let tabulate data : string =
-      let aux (l : string) log_line =
+      (* let first_idx, last_idx = (0, List.length data - 1) in *)
```
Comment (maiste) [inline web-ui/view/step.ml:569]: This can be removed.

File: web-ui/view/step.ml [hunk 7]
```diff
@@ -563,64 +579,73 @@ module Make (M : Git_forge_intf.Forge) = struct
...
+        | [] -> raise (Failure "Trying to take log_line from empty list (BUG)")
+        | s :: [] -> s
+        | _ :: l -> last l
+      in
+      match log_line with
+      | "" -> ""
+      | log_line ->
+          Dream.log "%S" log_line;
+          let split = Astring.String.cuts ~sep:"\r" log_line in
+          if List.length split > 1 then List.iter (Dream.log "%S") split;
```
Comment (maiste) [inline web-ui/view/step.ml:634]:
```suggestion
          if (List.compare_length_with split 1) >= 1 then List.iter (Dream.log "%S") split;
```
`List.compare_length_with` stops the computation at 1 iteration. It avoids unecessary computation.

File: web-ui/view/step.ml [hunk 8]
```diff
@@ -563,64 +579,73 @@ module Make (M : Git_forge_intf.Forge) = struct
...
     in
     let open Lwt.Infix in
     Dream.stream
       ~headers:[ ("Content-type", "text/html; charset=utf-8") ]
       (fun response_stream ->
         Dream.write response_stream header >>= fun () ->
-        let data' =
-          data |> Ansi.process ansi |> Astring.String.cuts ~sep:"\n" |> tabulate
-        in
+        let data' = data |> process_logs in
```
Comment (maiste) [inline web-ui/view/step.ml:648]:
```suggestion
        let data' = process_logs data in
```

File: web-ui/view/step.ml [hunk 9]
```diff
@@ -629,12 +654,7 @@ module Make (M : Git_forge_intf.Forge) = struct
               Dream.close response_stream
           | Ok (data, next) ->
               Dream.log "Fetching logs";
-              let data' =
-                data
-                |> Ansi.process ansi
-                |> Astring.String.cuts ~sep:"\n"
-                |> tabulate
-              in
+              let data' = data |> process_logs in
```
Comment (maiste) [inline web-ui/view/step.ml:657]:
```suggestion
              let data' = process_logs data in
```

**Comment by maiste** [comment]
git clone --recursive "https://github.com/maiste/ocaml-ci-experiment.git" -b "no-format" && cd "ocaml-ci-experiment" && git reset --hard a033e
cat > Dockerfile <<'END-OF-DOCKERFILE'
FROM ocaml/opam@sha:e4ce65485aed7b09b605f39643b46b86b913b54f5304e0dc125d815d18805887
# debian--4.14_opam-2.1
USER :1000
ENV CLICOLOR_FORCE=""
ENV OPAMCOLOR="always"
WORKDIR /src
RUN sudo ln -f /usr/bin/opam-.1 /usr/bin/opam
RUN opam init --reinit -ni
WORKDIR /src
RUN sudo chown opam /src
RUN cd ~/opam-repository && (git cat-file -e c64f99389df018a1aa5f82610c3fe74cc1cbbc8 || git fetch origin master) && git reset -q --hard 0c64f99389df018a1aa5f82610c3fe74cc1cbbc8 && git log --no-decorate -n1 --oneline && opam update -u
COPY --chown=:1000 test.opam ./
RUN opam pin add -yn test.dev './'
ENV DEPS="base-bigarray.base base-threads.base base-unix.base dune..6.1 ocaml.4.14.1 ocaml-base-compiler.4.14.1 ocaml-config.2 ocaml-options-vanilla.1"
ENV CI="true"
ENV OCAMLCI="true"
RUN opam update --depexts && opam install --cli=.1 --depext-only -y test.dev $DEPS
RUN opam install $DEPS
COPY --chown=:1000 . /src
RUN opam exec -- dune build @install @check @runtest && rm -rf _build

END-OF-DOCKERFILE
docker build .

**Review by maiste — CHANGES_REQUESTED** [review]
Only one change is required in the review here: `List.compare_length_with`. Otherwise, it can be merged on the master.

Working on Firefox.108.0.2 and Chromium.109.0.5414.74

**Pull request merged**